### DISCUSSION TITLE
Async handlers don't work if they begin with `let`/`const` or an `if` statement

### DISCUSSION THREAD

**Discussion by willcosgrove** [alpinejs/alpine#3118] [7. Bugs (archived)]
It looks like the issue is caused by wrapping the expression for safe right hand side evaluation:

https://github.com/alpinejs/alpine/blob/4df4e4f12aa929e69cde0cd0fde65ac624b8e5d8/packages/alpinejs/src/evaluator.js#L68-L77

On line 76, if the above conditions return true, the expression gets wrapped in `(() => ...)()` but that wrapping function is not async, and so any `await` calls will cause a syntax error.

**Comment by willcosgrove** [comment]
I was able to side step the bug by wrapping my own expression in an async function and calling it immediately.

```html
<button @click="(async () => { ... })()"></button>
```

**Reply by fweth** [reply]
Maybe related, I sometimes insert a newline for longer code blocks, if the `let` or `const` appears after some space or newline, it also seems to cause problems.  So
```html
<button @click="
  const foo = 3;
  console.log(foo);
">Click</button>
```
causes an error, but 
```html
<button @click="const foo = 3;
  console.log(foo);
">Click</button>
```
will not.

**Reply by ekwoka** [reply]
what's the error caused?

**Reply by fweth** [reply]
https://jsfiddle.net/p7Lr860a/6/ it's a long one, but it seems to be related to the missing `\s*` in the `let` expression.

**Comment by ekwoka** [comment]
This seems like it should be a simple PR to do, assuming there isn't any threads that pull off.

**Reply by SimoTod** [reply]
I think it's a safe change. The evaluator already resolves to an async function at the top level so making the intermediate function async shouldn't change anything. I don't think the OP is going to send a PR so feel free to submit one.

**Comment by fweth** [comment]
I don't understand the `async` problem well enough to do a pull request but I'd replace the expressions `/^[\n\s]*if.*\(.*\)/` resp. `/^(let|const)\s/` with `/^\s*if\s/` resp. `/^\s*(let|const)\s/` (I think `\s` also matches also `\n` and if there are no parantheses after the `if`, it will cause an JS error anyway and it's not up to Alpine.js to detect all JS syntax errors).

**Reply by SimoTod** [reply]
true, but your regexp assumes that there is a space after the if whilst some people just write `if(`. The parenthesis is not to detect a syntax error but to avoid false positives where it's just a variable starting with the letters `if`.
Peraphs `/^\s*if\s*(/` would do

**Reply by fweth** [reply]
Ah true!  So `/^\s*if\s*(/` or `/^\s*if[(,\s]/` then!

**Reply by ekwoka** [reply]
It's actually even easier then that: just add `async` to the IIFE 🤣

**Reply by fweth** [reply]
Wait, but aren't those two different problems?  I was just commenting on how to allow also `let`/`const` after a newline or similar constructs...

**Comment by ekwoka** [comment]
https://github.com/alpinejs/alpine/pull/3146

PR'd
